README
****************************************
tune-reporting-php
TUNE Reporting API SDK for PHP 5.3
Incorporate TUNE Reporting API services.
Update:  $Date: 2015-12-08 18:07:31 $
Version: 1.0.4
****************************************

=============
Overview
=============

PHP helper library for TUNE Reporting API services.

The utility focus of this SDK is upon the Advertiser Reporting endpoints.

Even though the the breadth of the Management API goes beyond just reports,
it is these endpoints that our customers primarily access.

The second goal of the SDKs is to assure that our customers’ developers
are using best practices in gathering reports in the most optimal way.

=============
Documentation
=============

Please see documentation here:

https://developers.mobileapptracking.com/tune-reporting-sdks/

=============
Requirements
=============

-------------
Prerequisites
-------------

* PHP >= 5.2.3
* The PHP JSON extension
* The PHP Curl extension

Examples requires setting within your `php.ini` parameter `date.timezone`:

```
[Date]
; Defines the default timezone used by the date functions
; http://php.net/date.timezone
date.timezone = "America/Los_Angeles"
```

-------------
Generate API Key
-------------

To use SDK, it requires you to generate API Key:
http://developers.mobileapptracking.com/generate-api-key/

=============
Installation
=============

-------------
Via Composer:
-------------

Add the following script within your `composer.json` file to include `tune-reporting-php`:

```
{
    "require": {
        "mobileapptracking/tune-reporting-php": "dev-master"
    },
    "require-dev": {
        "mobileapptracking/tune-reporting-php": "dev-master"
    },
    "repositories": [
        {
            "type": "vcs",
            "url": "git@github.com:MobileAppTracking/tune-reporting-php.git"
        }
    ],
}
```

-------------
Via ZIP file:
-------------

You can download the source code ZIP for `tune-reporting-php`:
https://github.com/MobileAppTracking/tune-reporting-php/archive/master.zip

Once you download the library, move the `tune-reporting-php` folder to your project
directory and then include the library file:

```php
    require '/path/to/tune-reporting-php/lib/TuneApi.php';
```

and you're good to go!

=============
Code Samples
=============

-------------
SDK Examples
-------------

Run the following script to view execution of all examples:

```
    make examples api_key=[API_KEY]
```

-------------
SDK Unittests
-------------

Run the following script to view execution of all unittests:

```
    rm -f composer.phar
    wget http://getcomposer.org/composer.phar
    php composer.phar install
    phpunit --version
    cd tests
    export API_KEY=demoadv
    phpunit --strict --stop-on-failure --configuration phpunit.xml
    cd ..
```

===========================
SDK Generated Documentation
===========================

SDK code is well commented and to see full documentation of its source using the provided Makefile commands that initiate code documentation generators.

----------------------------
SDK Documentation -- Doxygen
----------------------------

The following will generate Doxygen from PHP codebase:
http://en.wikipedia.org/wiki/Doxygen


This code documentation generation requires installation of Doxygen:
http://www.stack.nl/~dimitri/doxygen/index.html.

```
    make docs-doxygen
```

----------------------------------
SDK Documentation -- phpDocumentor
----------------------------------

The following will generate PhpDocumentor from PHP codebase:
http://en.wikipedia.org/wiki/PhpDocumentor

This code documentation generation requires installation of phpDocumentatior:
http://www.phpdoc.org/

```
    make docs-phpdoc
```

=============
License
=============

MIT License: http://opensource.org/licenses/MIT

=============
Reporting Issues
=============

We would love to hear your feedback.

Report issues using the Github Issue Tracker:
https://github.com/MobileAppTracking/tune-reporting-php/issues

or Email:
sdk@tune.com

CHANGES.md
## tune-reporting-php Changelog
=======================

Here you can see the full list of changes between each tune-reporting-php release.

Version 0.9.0
--------------

Internal release on October 06, 2014
* Tune Management API Client for PHP 5.3

Version 0.9.1
--------------

Internal release on October 12, 2014
* Tune Management API Advertise Reports

Version 0.9.2
--------------

Internal release on October 14, 2014
* Unittests and Examples

Version 0.9.3
--------------

Internal release on October 18, 2014
* Revised packaging

Version 0.9.5
--------------

Initial public release on October 22, 2014
* Field validation for parameter fields, sort, filter, group

Version 0.9.6
--------------

Beta public release on October 23, 2014
* Refactored layout of setup, examples, and tests.

Version 0.9.7
--------------

Beta public release on October 27, 2014
* Removed threading from fetch()
* Provided recommended values for exports()

Version 0.9.8
--------------

* Added methods fetch() and status() for all classes.

Version 0.9.9
--------------

Beta public release on October 30, 2014
* Created endpoint base classes.
* Removed bash scripts and provided Makefile instead.

Version 0.9.10
--------------

Beta public release on October 31, 2014
* Better handling of report worker.

Version 0.9.11
--------------

Beta public release on October 31, 2014
* Refactor: Renamed service classes.

ng of report worker.

Version 0.9.12
--------------

Beta public release on November 03, 2014
* Documentation ready to use phpDocumentor.

Version 0.9.13
--------------

Beta public release on November 05, 2014
* Sort changes for all Advertiser Reports.

Version 0.9.14
--------------

Beta public release on November 06, 2014
* Cohort endpoint changes.
* Actuals endpoint changes.

Version 0.9.16
--------------

Beta public release on Novenber 19, 2014
* Fixes to Cohort and Retention reports

Version 0.9.20
--------------

Beta public release on December 10, 2014
* Refactor to reflect new naming from 'tune-api-php' to 'tune-reporting-php'

Version 0.9.23
--------------

Beta public release on December 19, 2014
* SDK Configuration

Version 0.9.24
--------------

Beta public release on December 19, 2014
* /advertiser/stats/ltv is now considered as Value reporting.

Version 0.9.25
--------------

Beta public release on December 24, 2014
* Refactor Log and Cohort classes
* TUNE Management API request contains SDK type and version.

Version 0.9.26
--------------

Beta public release on December 30, 2014
* TUNE Management API request contains 'sdk' and 'ver'.

Version 0.9.27
--------------

Beta public release on December 31, 2014
* TUNE Reporting SDK set API_KEY.

Version 0.9.28
--------------

Beta public release on January 03, 2015
* Provide toJson() within response.
* Created 'config' directory to hold TUNE Reporting SDK configuration.

Version 0.9.29
--------------

Beta public release on January 05, 2015
* Provide both authentication approaches: api_key and session_token.
* New class ```SessionAuthentication``` for generating session_token.

Version 1.0.0
--------------

First production release on April 09, 2015


Version 1.0.4
--------------

Modification of TUNE Reporting endpoint /advertiser/stats/retention.
AdvertiserReportCohortRetention has new parameter retention_measure.

examples/ExampleClicks.php
<?php
/**
 * ExampleClicks.php
 *
 * Copyright (c) 2014 Tune, Inc
 * All rights reserved.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 *
 * PHP Version 5.3
 *
 * @category  Tune
 * @package   Tune_API_PHP
 * @author    Jeff Tanner <jefft@tune.com>
 * @copyright 2014 Tune (http://www.tune.com)
 * @license   http://opensource.org/licenses/MIT The MIT License (MIT)
 * @version   0.9.8
 * @link      https://developers.mobileapptracking.com Tune Developer Community @endlink
 *
 */

require_once dirname(__FILE__) . "/../src/TuneApi.php";

use Tune\Management\Api\Advertiser\Stats\Clicks;
use Tune\Management\Shared\Service\TuneManagementBase;
use Tune\Management\Shared\Reports\ReportReaderCSV;
use Tune\Management\Shared\Reports\ReportReaderJSON;

global $argc, $argv;

/**
 * Class ExampleClicks
 *
 * Using Tune\Management\Api\Advertiser\Stats\Clicks
 */
class ExampleClicks
{
    /**
     * Constructor that prevents a default instance of this class from being created.
     */
    private function __construct()
    {

    }

    /**
     * Execute example
     *
     * @param string $api_key MobileAppTracking API Key
     *
     * @throws \InvalidArgumentException
     * @throws \Exception
     */
    public static function run($api_key)
    {
        // api_key
        if (!is_string($api_key) || empty($api_key)) {
            throw new \InvalidArgumentException("Parameter 'api_key' is not defined.");
        }

        $default_date_timezone = ini_get('date.timezone');
        if (is_string($default_date_timezone) && !empty($default_date_timezone)) {
            echo "======================================================" . PHP_EOL;
            echo " Default timezone used: '{$default_date_timezone}'." . PHP_EOL;
            echo "======================================================" . PHP_EOL;
        } else {
            throw new \RuntimeException(
                "It is not safe to rely on the system's timezone settings. "
                . "You are *required* to use the date.timezone setting or "
                . "the date_default_timezone_set() function."
            );
        }

        echo "=========================================================" . PHP_EOL;
        echo "= Tune Management API Advertiser Reports Logs Clicks    =" . PHP_EOL;
        echo "=========================================================" . PHP_EOL;

        try {
            $yesterday      = date('Y-m-d', strtotime("-1 days"));
            $start_date     = "{$yesterday} 00:00:00";
            $end_date       = "{$yesterday} 23:59:59";

            $clicks = new Clicks($api_key, $validate_fields = true);

            echo "======================================================" . PHP_EOL;
            echo " Fields of Advertiser Logs Clicks records.           " . PHP_EOL;
            echo "======================================================" . PHP_EOL;
            $response = $clicks->fields();
            echo print_r($response, true) . PHP_EOL;

            echo "======================================================" . PHP_EOL;
            echo " Count Advertiser Logs Clicks records.                " . PHP_EOL;
            echo "======================================================" . PHP_EOL;
            $response = $clicks->count(
                $start_date,
                $end_date,
                $filter              = null,
                $response_timezone   = "America/Los_Angeles"
            );

            echo "= Response:" . PHP_EOL;
            echo print_r($response, true) . PHP_EOL;

            if ($response->getHttpCode() != 200) {
                throw new \Exception(
                    sprintf("Failed: %d: %s", $response->getHttpCode(), print_r($response->getErrors()))
                );
            }

            echo "= Count:" . $response->getData() . PHP_EOL;

            echo "======================================================" . PHP_EOL;
            echo " Find Advertiser Logs Clicks records.                 " . PHP_EOL;
            echo "======================================================" . PHP_EOL;

            $response = $clicks->find(
                $start_date,
                $end_date,
                $filter              = null,
                $fields              = $clicks->fields(TuneManagementBase::TUNE_FIELDS_DEFAULT | TuneManagementBase::TUNE_FIELDS_MINIMAL),
                $limit               = 5,
                $page                = null,
                $sort                = array("created" => "DESC"),
                $response_timezone   = "America/Los_Angeles"
            );

            echo "= Response:" . PHP_EOL;
            echo print_r($response, true) . PHP_EOL;

            if ($response->getHttpCode() != 200) {
                throw new \Exception(
                    sprintf("Failed: %d: %s", $response->getHttpCode(), print_r($response->getErrors()))
                );
            }

            echo "==========================================================" . PHP_EOL;
            echo " Request Advertiser Logs Clicks CSV report for export.    " . PHP_EOL;
            echo "==========================================================" . PHP_EOL;

            $response = $clicks->export(
                $start_date,
                $end_date,
                $filter              = null,
                $fields              = $clicks->fields(TuneManagementBase::TUNE_FIELDS_RECOMMENDED),
                $format              = "csv",
                $response_timezone   = "America/Los_Angeles"
            );

            echo "= Response:" . PHP_EOL;
            echo print_r($response, true) . PHP_EOL;

            if ($response->getHttpCode() != 200) {
                throw new \Exception(
                    sprintf("Failed: %d: %s", $response->getHttpCode(), print_r($response->getErrors()))
                );
            }

            $job_id = Clicks::parseResponseReportJobId($response);
            echo "= CSV Job ID: {$job_id}" . PHP_EOL;

            echo "=======================================================" . PHP_EOL;
            echo " Fetching Advertiser Logs Clicks CSV report.           " . PHP_EOL;
            echo "=======================================================" . PHP_EOL;

            $response = $clicks->fetch(
                $job_id,
                $verbose = true,
                $sleep = 10
            );

            $report_url = Clicks::parseResponseReportUrl($response);
            echo "= CSV Report URL: {$report_url}" . PHP_EOL;

            echo "======================================================" . PHP_EOL;
            echo " Read Clicks CSV report and pretty print 5 lines.     " . PHP_EOL;
            echo "======================================================" . PHP_EOL;
            $csv_report_reader = new ReportReaderCSV(
                $report_url
            );

            $csv_report_reader->read();
            $csv_report_reader->prettyPrint($limit = 5);

            echo "======================================================" . PHP_EOL;
            echo " Request Advertiser Clicks JSON report for export.    " . PHP_EOL;
            echo "======================================================" . PHP_EOL;

            $response = $clicks->export(
                $start_date,
                $end_date,
                $filter              = null,
                $fields              = $clicks->fields(TuneManagementBase::TUNE_FIELDS_RECOMMENDED),
                $format              = "json",
                $response_timezone   = "America/Los_Angeles"
            );

            echo "= Response:" . PHP_EOL;
            echo print_r($response, true) . PHP_EOL;

            if ($response->getHttpCode() != 200) {
                throw new \Exception(
                    sprintf("Failed: %d: %s", $response->getHttpCode(), print_r($response->getErrors()))
                );
            }

            $job_id = Clicks::parseResponseReportJobId($response);
            echo "= JSON Job ID: {$job_id}" . PHP_EOL;

            echo "========================================================" . PHP_EOL;
            echo " Fetching Advertiser Logs Clicks JSON report            " . PHP_EOL;
            echo "========================================================" . PHP_EOL;

            $response = $clicks->fetch(
                $job_id,
                $verbose = true,
                $sleep = 10
            );

            $report_url = Clicks::parseResponseReportUrl($response);
            echo "= JSON Report URL: {$report_url}" . PHP_EOL;

            echo "========================================================" . PHP_EOL;
            echo " Read Clicks JSON report and pretty print 5 lines.      " . PHP_EOL;
            echo "========================================================" . PHP_EOL;

            $json_report_reader = new ReportReaderJSON(
                $report_url
            );

            $json_report_reader->read();
            $json_report_reader->prettyPrint($limit = 5);

            echo "======================================================" . PHP_EOL;
            echo " End Example                                          " . PHP_EOL;
            echo "======================================================" . PHP_EOL;
            echo PHP_EOL;

        } catch (\Exception $ex) {
            throw $ex;
        }
    }
}

/**
 * Example request API_KEY
 */
if (count($argv) == 1) {
    echo sprintf("%s [api_key]", $argv[0]) . PHP_EOL;
    exit;
}

$api_key = $argv[1];

ExampleClicks::run($api_key);
